### PR TITLE
Cargo integration test now checks mean price of random crates

### PR DESCRIPTION
**Pull request by riccardi48**
## About the PR

Cargo integration test now checks mean price of random crates. Also removed FunCrateGambling (Lotto crate) from the ignored list as it now passes.

## Why / Balance

Any changes to the weights of expensive items in the lotto crate might change the mean value more than expected. If the mean value is more than the cost it becomes an infinite money printer.

## Technical details

Gets the `EntityTableSystem.AverageSpawns()` dict which lists each item and their expected amount spawned. Times by its prices and summed gets the mean price of a spawning cargo product.
It skips if the crate has no randomness and checks it normally.
Currently it only gets the price of all the items spawned not the container itself so might be $75 off or so.

## Media

Forcing a fail for image. Mean price of lotto crate is ~$5005 exluding the price of the crate
Code_xgt7SDvtSB

## Requirements

- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an in-game showcase.

### CODE REVIEW

**Comment by riccardi48** [comment]
EnsureValidRandomSpecies("Vulpkanin") HeisenTest for fail. The cargo test passed.

File: Content.IntegrationTests/Tests/CargoTest.cs [hunk 1]
```diff
@@ -1,134 +1,165 @@
...
-        var testMap = await pair.CreateTestMap();
-
-        var entManager = server.ResolveDependency<IEntityManager>();
-        var protoManager = server.ResolveDependency<IPrototypeManager>();
-        var pricing = server.ResolveDependency<IEntitySystemManager>().GetEntitySystem<PricingSystem>();
-
-        await server.WaitAssertion(() =>
+        await Server.WaitAssertion(() =>
         {
             Assert.Multiple(() =>
```
Comment (Tayrtahn) [inline Content.IntegrationTests/Tests/CargoTest.cs]:
```suggestion
            using (Assert.EnterMultipleScope())
```

File: Content.IntegrationTests/Tests/CargoTest.cs [hunk 2]
```diff
@@ -1,134 +1,165 @@
...
 
 [TestFixture]
 public sealed class CargoTest : GameTest
 {
-    private static readonly HashSet<ProtoId<CargoProductPrototype>> Ignored =
-    [
-        // This is ignored because it is explicitly intended to be able to sell for more than it costs.
-        new("FunCrateGambling")
-    ];
+    private static readonly HashSet<ProtoId<CargoProductPrototype>> Ignored = [];
```
Comment (Tayrtahn) [inline Content.IntegrationTests/Tests/CargoTest.cs]: Provide some room for forks to add entries without conflicts, and add some documentation.
```suggestion
    /// <summary>
    /// <see cref="NoCargoOrderArbitrage"/> will ignore all <see cref="CargoProductPrototype"/>s listed here.
    /// </summary>
    private static readonly HashSet<ProtoId<CargoProductPrototype>> Ignored =
    [

    ];
```

**Review by Tayrtahn — CHANGES_REQUESTED** [review]

**Review by Tayrtahn — APPROVED** [review]

File: Content.IntegrationTests/Tests/CargoTest.cs [hunk 3]
```diff
@@ -1,134 +1,171 @@
...
+                                Is.AtMost(proto.Cost),
+                                $"Found arbitrage on {proto.ID} cargo product!  Cost is {proto.Cost} but mean sell price is {price}!"
+                            );
+                            continue;
+                        }
+                    }
 
-                    var ent = entManager.SpawnEntity(proto.Product, testMap.MapCoords);
-                    var price = pricing.GetPrice(ent);
+                    ent = SSpawnAtPosition(proto.Product, coordinates);
```
Comment (slarticodefast) [inline Content.IntegrationTests/Tests/CargoTest.cs:74]: Aren't you counting double in this case?
First you add the average price of the random entities,
and then add the price of the total product, which again spawns some of the random entities and includes their price.

File: Content.IntegrationTests/Tests/CargoTest.cs [hunk 4]
```diff
@@ -1,134 +1,171 @@
...
+                foreach (var proto in SProtoMan.EnumeratePrototypes<CargoProductPrototype>())
                 {
                     if (Ignored.Contains(proto.ID))
                         continue;
+                    var entProto = SProtoMan.Index<EntityPrototype>(proto.Product);
+                    double price = 0;
+                    EntityUid ent;
+                    if (entProto.TryGetComponent<EntityTableContainerFillComponent>(out var fill, _sCompFact))
+                    {
+                        var averageSpawns = _sTableSystem.AverageSpawns(fill.Containers.First().Value);
```
Comment (slarticodefast) [inline Content.IntegrationTests/Tests/CargoTest.cs]: Shouldn't this count all containers?

File: Content.IntegrationTests/Tests/CargoTest.cs (hunk 3 above)
Comment (riccardi48) [reply Content.IntegrationTests/Tests/CargoTest.cs:74]: The continue above means if the price was found using the average it doesn't spawn the overall container. I'll add a comment to make that clearer.

File: Content.IntegrationTests/Tests/CargoTest.cs (hunk 4 above)
Comment (riccardi48) [reply Content.IntegrationTests/Tests/CargoTest.cs]: Not all crates are have tables. Some, like the jani trolly, are just single entities. For those that still have tables, I assume spawning each component and pricing would be slower than spawning the whole table and getting the price at once.

Comment (riccardi48) [reply Content.IntegrationTests/Tests/CargoTest.cs]: Reading this back, I think you meant shouldn't it check all containers. Containers only achieve randomness using the EntityTableContainerFillComponent. So if it doesn't use that component it must be deterministic and therefore skipped.

**Comment by riccardi48** [comment]
This is finally, for real this time, ready for review.

File: Content.IntegrationTests/Tests/CargoTest.cs [hunk 5]
```diff
@@ -35,34 +35,61 @@ public sealed class CargoTest : GameTest
...
+                        // Randomness will lead to non integer expected values, if all the expected values are integers then we skip
+                        // Compares against epsilon in case of any floating point stuff
+                        if (!averageSpawns.All(item => Math.Abs(item.Item2 % 1) <= Double.Epsilon * 100))
+                        {
```
Comment (whatston3) [inline Content.IntegrationTests/Tests/CargoTest.cs]: I'm not sure that this condition works as you expect.
Suppose you get a value pack of produce with a random fill.
It contains 2-4 grapes with probability 1, 3-7 oranges with probability 0.4, and 3-5 apples with probability 0.25.
That's a random fill, whose mean value should be respected.
However, it contains an average spawn count of 3 grapes, 2 oranges, and 1 apple, which would fail this check, no?  Those are all distributions with whole means, though they have non-zero variance.
I don't think you should skip this [when this condition is true].

File: Content.IntegrationTests/Tests/CargoTest.cs [hunk 6]
```diff
@@ -35,34 +35,61 @@ public sealed class CargoTest : GameTest
...
+                            foreach (var item in averageSpawns)
+                            {
+                                ent = SSpawnAtPosition(item.spawn, coordinates);
+                                price += _sPricing.GetPrice(ent) * item.Item2;
+                                SDeleteNow(ent);
+                            }
+                            // Price of container is not included right now
```
Comment (whatston3) [inline Content.IntegrationTests/Tests/CargoTest.cs]: Spawn the item anyways, check only the base, not the contents.
Should work, no?
```cs
// TODO: Spawn your item anyways (as ent), then only get the value of the base object.
price += _sPricing.GetPrice(ent, includeContents: false);
```

**Review by whatston3 — CHANGES_REQUESTED** [review]
A few points, but it looks good!

File: Content.IntegrationTests/Tests/CargoTest.cs (hunk 5 above)
Comment (riccardi48) [reply Content.IntegrationTests/Tests/CargoTest.cs]: I agree this is an edge case which is missed, though, I'm not sure anything in the game currently has it. The options are to either, use a different, much more complex, way to determine if the fill is random, or spawn in every one. My concern is that there are way more deterministic fills than random and spawning the contents of each separately might end up taking a lot of time. I'll do a check, when I can, if this is actually much of a diffrence in time taken .The best solution imo would be for the table system to have a function added to the interface to check if a table has variance or not.

File: Content.IntegrationTests/Tests/CargoTest.cs (hunk 6 above)
Comment (riccardi48) [reply Content.IntegrationTests/Tests/CargoTest.cs]: I didn't realise that argument existed, this is perfect for checking the crate price and has been added.

File: Content.IntegrationTests/Tests/CargoTest.cs (hunk 4 above)
Comment (SnappingOpossum) [reply Content.IntegrationTests/Tests/CargoTest.cs]: EntityTableContainerFill can fill multiple containers though, and if at some point a purchasable has multiple randomly filled containers those should be accounted for.

Comment (riccardi48) [reply Content.IntegrationTests/Tests/CargoTest.cs]: yeah, I super missunderstood the original review. I have fixed this now to check through each.

File: Content.IntegrationTests/Tests/CargoTest.cs (hunk 5 above)
Comment (riccardi48) [reply Content.IntegrationTests/Tests/CargoTest.cs]: I've checked and the time is not that much. I've just changed it to not do this check at all and just run through each.
